invokeai/app/services/object_serializer/object_serializer_disk.py

```diff
@@ -0,0 +1,95 @@
+
+if TYPE_CHECKING:
+    from invokeai.app.services.invoker import Invoker
+
+
+T = TypeVar("T")
+
+
+@dataclass
+class DeleteAllResult:
+    deleted_count: int
+    freed_space_bytes: float
+
+
+class ObjectSerializerDisk(ObjectSerializerBase[T]):
+    """Provides a disk-backed storage for arbitrary python objects.
+
+    :param output_folder: The folder where the objects will be stored
+    :param delete_on_startup: If True, all objects in the output folder will be deleted on startup
+    """
+
+    def __init__(self, output_dir: Path, delete_on_startup: bool = False):
+        super().__init__()
+        self._output_dir = output_dir
+        self._output_dir.mkdir(parents=True, exist_ok=True)
+        self._delete_on_startup = delete_on_startup
+        self.__obj_class_name: Optional[str] = None
+
+    def start(self, invoker: "Invoker") -> None:
+        self._invoker = invoker
+
+        if self._delete_on_startup:
+            delete_all_result = self._delete_all()
+            if delete_all_result.deleted_count > 0:
+                freed_space_in_mb = round(delete_all_result.freed_space_bytes / 1024 / 1024, 2)
+                self._invoker.services.logger.info(
+                    f"Deleted {delete_all_result.deleted_count} {self._obj_class_name} files (freed {freed_space_in_mb}MB)"
+                )
+
+    def load(self, name: str) -> T:
+        file_path = self._get_path(name)
+        try:
+            return torch.load(file_path)  # pyright: ignore [reportUnknownMemberType]
+        except FileNotFoundError as e:
+            raise ObjectNotFoundError(name) from e
+
+    def save(self, obj: T) -> str:
+        name = self._new_name()
+        file_path = self._get_path(name)
+        torch.save(obj, file_path)  # pyright: ignore [reportUnknownMemberType]
+        return name
+
+    def delete(self, name: str) -> None:
+        file_path = self._get_path(name)
+        file_path.unlink()
+
+    @property
+    def _obj_class_name(self) -> str:
+        if not self.__obj_class_name:
+            # `__orig_class__` is not available in the constructor for some technical, undoubtedly very pythonic reason
+            self.__obj_class_name = typing.get_args(self.__orig_class__)[0].__name__  # pyright: ignore [reportUnknownMemberType, reportGeneralTypeIssues]
+        return self.__obj_class_name
+
+    def _get_path(self, name: str) -> Path:
+        return self._output_dir / name
+
+    def _new_name(self) -> str:
+        return f"{self._obj_class_name}_{uuid_string()}"
+
+    def _delete_all(self) -> DeleteAllResult:
+        """
+        Deletes all objects from disk.
+        """
+
+        # We could try using a temporary directory here, but they aren't cleared in the event of a crash, so we'd have
+        # to manually clear them on startup anyways. This is a bit simpler and more reliable.
+
+        deleted_count = 0
+        freed_space = 0
+        for file in Path(self._output_dir).glob("*"):
```

Comment:
This `glob("*")` makes me slightly nervous. Maybe I'm just being paranoid... but I'd hate for someone to accidentally wipe a directory by passing in the wrong `output_dir`.

Some options I can think of to reduce risk:
1. Use a temp directory. I saw your comment above, but I think temp directories should _almost_ always be cleaned up correctly. The only exception I can think of is if a daemon thread is killed without proper cleanup.
2. Always use an output_dir in the system's temp dir. Before deleting, check to make sure that the path is in fact in the temp dir.
3. Use a file extension, and check that we are only deleting files with that file extension.

Reply:
Ok, I've figured out a solution with `tempfile.TemporaryDirectory`. From the commit message:

Replace `delete_on_startup: bool` & associated logic with `ephemeral: bool` and `TemporaryDirectory`.

The temp dir is created inside of `output_dir`. For example, if `output_dir` is `invokeai/outputs/tensors/`, then the temp dir might be `invokeai/outputs/tensors/tmpvj35ht7b/`.

The temp dir is cleaned up when the service is stopped, or when it is GC'd if not properly stopped.

In the event of a catastrophic crash where the temp files are not cleaned up, the user can delete the tempdir themselves.

This situation may not occur in normal use, but if you kill the process, python cannot clean up the temp dir itself. This includes running the app in a debugger and killing the debugger process - something I do relatively often.

Tests updated.